File: src/test/java/com/raychenon/leetcode/FloodFillTest.java
package com.raychenon.leetcode;

import com.raychenon.kotlin.ext.ArrayExtKt;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;


/**
 * User: raychenon
 * Date: 2019-09-18
 * https://leetcode.com/problems/flood-fill/
 */
public class FloodFillTest {

    private static FloodFill floodfill;

    @BeforeClass
    public static void setUp() {
        floodfill = new FloodFill();
    }

    @Test
    public void fillMatrix3x3Test() {
        int[][] image = new int[][]{
                {1, 1, 1},
                {1, 1, 0},
                {1, 0, 1}
        };
        int[][] result = new int[][]{
                {2, 2, 2},
                {2, 2, 0},
                {2, 0, 1}
        };

        compareFloodFillMethods(result, image, 1, 1, 2);
    }

    @Test
    public void fillMatrix4x3Test() {
        int[][] image = new int[][]{
                {1, 1, 1, 1},
                {1, 1, 0, 2},
                {1, 0, 1, 1}
        };
        int[][] result = new int[][]{
                {2, 2, 2, 2},
                {2, 2, 0, 2},
                {2, 0, 1, 1}
        };

        compareFloodFillMethods(result, image, 1, 1, 2);
    }

    @Test
    public void sameColorTest() {
        int[][] image = new int[][]{
                {1, 1, 1},
                {1, 1, 0},
                {1, 0, 1}
        };
        int[][] result = new int[][]{
                {1, 1, 1},
                {1, 1, 0},
                {1, 0, 1}
        };

        compareFloodFillMethods(result, image, 1, 1, 1);
    }


    private void compareFloodFillMethods(int[][] expected, int[][] image, int row, int col, int newColor) {
        // image parameter is modified by reference, so pass a copy
        Assert.assertArrayEquals(expected, floodfill.floodFillIterative(ArrayExtKt.copyDeep(image), row, col, newColor));
        Assert.assertArrayEquals(expected, floodfill.floodFill(ArrayExtKt.copyDeep(image), row, col, newColor));
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/FloodFillTest.kt
package com.raychenon.kotlin.leetcode

import org.junit.Assert
import org.junit.Test

/**
 * User: raychenon
 * Date: 2019-09-18
 * https://leetcode.com/problems/flood-fill
 */
class FloodFillTest {

    @Test
    fun fillMatrix3x3Test() {
        val image = arrayOf(
            intArrayOf(1, 1, 1),
            intArrayOf(1, 1, 0),
            intArrayOf(1, 0, 1)
        )
        val result = arrayOf(
            intArrayOf(2, 2, 2),
            intArrayOf(2, 2, 0),
            intArrayOf(2, 0, 1)
        )

        compareFloodFillMethods(result, image, 1, 1, 2)
    }

    @Test
    fun fillMatrix4x3Test() {
        val image = arrayOf(
            intArrayOf(1, 1, 1, 1),
            intArrayOf(1, 1, 0, 2),
            intArrayOf(1, 0, 1, 1)
        )
        val result = arrayOf(
            intArrayOf(2, 2, 2, 2),
            intArrayOf(2, 2, 0, 2),
            intArrayOf(2, 0, 1, 1)
        )

        compareFloodFillMethods(result, image, 1, 1, 2)
    }

    @Test
    fun sameColorTest() {
        val image = arrayOf(
            intArrayOf(1, 1, 1),
            intArrayOf(1, 1, 0),
            intArrayOf(1, 0, 1)
        )
        val result = arrayOf(
            intArrayOf(1, 1, 1),
            intArrayOf(1, 1, 0),
            intArrayOf(1, 0, 1)
        )
        compareFloodFillMethods(result, image, 1, 1, 1)
    }


    fun compareFloodFillMethods(expected: Array<IntArray>, image: Array<IntArray>, row: Int, col: Int, newColor: Int) {
        // image parameter is modified by reference, so pass a copy
        Assert.assertArrayEquals(expected, FloodFill.floodFillRec(image.clone(), row, col, newColor))
        Assert.assertArrayEquals(expected, FloodFill.floodFill(image.clone(), row, col, newColor))
    }
}